librz/type/parser/c_cpp_parser.c

```diff
@@ -0,0 +1,355 @@
+		// Print the syntax tree as an S-expression.
+		char *string = ts_node_string(root_node);
+		parser_debug(state, "Syntax tree: %s\n", string);
+		free(string);
+	}
+
+	// At first step we should handle defines
+	// #define
+	// #if / #ifdef
+	// #else
+	// #endif
+	// After that, we should process include files and #error/#warning/#pragma
+	// Temporarily we could just run preprocessing step using tccpp code
+	//
+	// And only after that - run the normal C/C++ syntax parsing
+
+	// Filter types function prototypes and start parsing
+	int i = 0, result = 0;
+	for (i = 0; i < root_node_child_count; i++) {
+		TSNode child = ts_node_named_child(root_node, i);
+		// We skip ";" or "," - empty expressions
+		const char *node_type = ts_node_type(child);
+		if (!strcmp(node_type, "expression_statement")) {
+			continue;
+		}
+		parser_debug(state, "Processing %d child...\n", i);
+		result += parse_type_nodes_save(state, child, code);
+	}
+
+	// If there were errors during the parser then the result is different from 0
+	if (result) {
+		const char *error_msgs = rz_strbuf_drain_nofree(state->errors);
+		eprintf("Errors:\n");
+		eprintf("%s", error_msgs);
+		const char *warning_msgs = rz_strbuf_drain_nofree(state->warnings);
+		eprintf("Warnings:\n");
+		eprintf("%s", warning_msgs);
+		if (error_msg) {
+			*error_msg = strdup(error_msgs);
+		}
+	}
+	if (state->verbose) {
+		const char *debug_msgs = rz_strbuf_drain_nofree(state->debug);
+		eprintf("%s", debug_msgs);
+	}
+
+	// After everything parsed, we should preserve the base type database
+	c_parser_state_free_keep_ht(state);
+	ts_tree_delete(tree);
+	ts_parser_delete(parser);
+	return result;
+}
+
+/**
+ * \brief Parses the C type string reusing the existing parser state
+ *
+ * \param parser RzTypeParser instance
+ * \param code The C type itself
+ * \param error_msg A pointer where all error messages will be stored
+ */
+RZ_API int rz_type_parse_string_stateless(RzTypeParser *parser, const char *code, char **error_msg) {
+	return type_parse_string(parser->state, code, error_msg);
+}
+
+/**
+ * \brief Parses the C types file reusing the existing parser state
+ *
+ * \param parser RzTypeParser instance
+ * \param path The path to the C file to parse
+ * \param dir The directory where the C file is located
+ * \param error_msg A pointer where all error messages will be stored
+ */
+RZ_API int rz_type_parse_file_stateless(RzTypeParser *parser, const char *path, const char *dir, char **error_msg) {
+	size_t read_bytes = 0;
+	const char *source_code = rz_file_slurp(path, &read_bytes);
+	if (!source_code || !read_bytes) {
+		return -1;
+	}
+	ut64 file_size = rz_file_size(path);
+	RZ_LOG_DEBUG("File size is %" PFMT64d " bytes, read %zu bytes\n", file_size, read_bytes);
+	return rz_type_parse_string_stateless(parser, source_code, error_msg);
+}
+
+/**
+ * \brief Parses the C types file creating the new parser state
+ *
+ * \param typedb RzTypeDB instance
```

Comment:
What does `creating the new parser state` mean? Also it should be made more clear, maybe even through the function names, that these functions actually also save the types in the `RzTypeDB` and thus mutate it, instead of just parsing.

Reply:
Creating new parser state means:
- Create new hashtable for base types
- Create new hashtable for callables
- Create new hashtable for forward declarations